core/src/main/java/com/scalar/db/transaction/consensuscommit/ConsensusCommitOperationChecker.java

```diff
@@ -0,0 +1,253 @@
+    TransactionTableMetadata metadata =
+        getTransactionTableMetadata(transactionTableMetadataManager, scan);
+
+    // Skip checks if including metadata columns is enabled
+    if (!isIncludeMetadataEnabled) {
+      // Check projections
+      for (String column : scan.getProjections()) {
+        if (metadata.getTransactionMetaColumnNames().contains(column)) {
+          throw new IllegalArgumentException(
+              CoreError
+                  .CONSENSUS_COMMIT_SPECIFYING_TRANSACTION_METADATA_COLUMNS_IN_PROJECTION_NOT_ALLOWED
+                  .buildMessage(scan.forFullTableName().get(), column));
+        }
+      }
+
+      // Check conditions
+      for (Selection.Conjunction conjunction : scan.getConjunctions()) {
+        for (ConditionalExpression condition : conjunction.getConditions()) {
+          String column = condition.getColumn().getName();
+          if (metadata.getTransactionMetaColumnNames().contains(column)) {
+            throw new IllegalArgumentException(
+                CoreError
+                    .CONSENSUS_COMMIT_CONDITION_NOT_ALLOWED_TO_TARGET_TRANSACTION_METADATA_COLUMNS
+                    .buildMessage(scan.forFullTableName().get(), column));
+          }
+        }
+      }
+
+      // Check orderings
+      for (Scan.Ordering ordering : scan.getOrderings()) {
+        String column = ordering.getColumnName();
+        if (metadata.getTransactionMetaColumnNames().contains(column)) {
+          throw new IllegalArgumentException(
+              CoreError
+                  .CONSENSUS_COMMIT_SPECIFYING_TRANSACTION_METADATA_COLUMNS_IN_ORDERING_NOT_ALLOWED
+                  .buildMessage(scan.forFullTableName().get(), column));
+        }
+      }
+    }
+
+    // Additional checks for SERIALIZABLE isolation level
+    if (context.isolation == Isolation.SERIALIZABLE) {
+      // Don't allow index scans
+      TableMetadata tableMetadata = metadata.getTableMetadata();
+      if (ScalarDbUtils.isSecondaryIndexSpecified(scan, tableMetadata)) {
+        // If the index column is part of the primary key, it's allowed
+        String indexKeyColumnName = scan.getPartitionKey().getColumns().get(0).getName();
+        if (!tableMetadata.getPartitionKeyNames().contains(indexKeyColumnName)
+            && !tableMetadata.getClusteringKeyNames().contains(indexKeyColumnName)) {
+          throw new IllegalArgumentException(
+              CoreError.CONSENSUS_COMMIT_INDEX_SCAN_NOT_ALLOWED_IN_SERIALIZABLE.buildMessage());
+        }
+      }
+
+      // If the scan is a cross-partition scan (ScanAll), don't allow conditions on indexed columns
+      if (scan instanceof ScanAll) {
+        for (Selection.Conjunction conjunction : scan.getConjunctions()) {
+          for (ConditionalExpression condition : conjunction.getConditions()) {
+            String column = condition.getColumn().getName();
+            if (metadata.getSecondaryIndexNames().contains(column)) {
```

Comment:
Sorry, one more small question. Is this prohibited even if this column is the first primary key column?